zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/FederatedZkClient.java

```diff
@@ -20,344 +20,487 @@
   public boolean waitUntilConnected(long time, TimeUnit timeUnit) {
+    throwUnsupportedOperationException();
     return false;
   }
 
   @Override
   public String getServers() {
+    throwUnsupportedOperationException();
     return null;
   }
 
   @Override
   public long getSessionId() {
-    return 0;
+    // Session-aware is unsupported.
+    throwUnsupportedOperationException();
+    return 0L;
   }
 
   @Override
   public void close() {
+    if (isClosed()) {
+      return;
+    }
 
+    LOG.info("Closing {}.", FEDERATED_ZK_CLIENT);
+
+    synchronized (_zkRealmToZkClientMap) {
+      _zkRealmToZkClientMap.values().forEach(ZkClient::close);
+      _zkRealmToZkClientMap.clear();
+      _isClosed = true;
+    }
   }
 
   @Override
   public boolean isClosed() {
-    return false;
+    return _isClosed;
   }
 
   @Override
   public byte[] serialize(Object data, String path) {
-    return new byte[0];
+    return getZkClient(path).serialize(data, path);
   }
 
   @Override
   public <T> T deserialize(byte[] data, String path) {
-    return null;
+    return getZkClient(path).deserialize(data, path);
   }
 
   @Override
   public void setZkSerializer(ZkSerializer zkSerializer) {
-
+    _pathBasedZkSerializer = new BasicZkSerializer(zkSerializer);
+    _zkRealmToZkClientMap.values()
+        .forEach(zkClient -> zkClient.setZkSerializer(_pathBasedZkSerializer));
   }
 
   @Override
   public void setZkSerializer(PathBasedZkSerializer zkSerializer) {
-
+    _pathBasedZkSerializer = zkSerializer;
+    _zkRealmToZkClientMap.values().forEach(zkClient -> zkClient.setZkSerializer(zkSerializer));
   }
 
   @Override
   public PathBasedZkSerializer getZkSerializer() {
-    return null;
+    return _pathBasedZkSerializer;
+  }
+
+  private String create(final String path, final Object dataObject, final List<ACL> acl,
+      final CreateMode mode, final String expectedSessionId) {
+    if (mode.isEphemeral()) {
+      throwUnsupportedOperationException();
+    }
+
+    // Create mode is not session-aware, so the node does not have to be created
+    // by the expectedSessionId.
+    return getZkClient(path).create(path, dataObject, acl, mode);
+  }
+
+  private ZkClient getZkClient(String path) {
+    // If FederatedZkClient is closed, should not return ZkClient.
+    if (isClosed()) {
+      throw new IllegalStateException(FEDERATED_ZK_CLIENT + " is closed!");
+    }
+
+    String zkRealm = getZkRealm(path);
+    if (!_zkRealmToZkClientMap.containsKey(zkRealm)) {
+      // Synchronized to avoid creating duplicate ZkClient for the same ZkRealm.
+      synchronized (_zkRealmToZkClientMap) {
+        if (!_zkRealmToZkClientMap.containsKey(zkRealm)) {
+          _zkRealmToZkClientMap.put(zkRealm, createZkClient(zkRealm));
+        }
+      }
+    }
+
+    return _zkRealmToZkClientMap.get(zkRealm);
+  }
+
+  private String getZkRealm(String path) {
+    String zkRealm;
+    try {
+      zkRealm = _metadataStoreRoutingData.getMetadataStoreRealm(path);
```

Comment:
Is this _metadataStoreRoutingData query thread safe? 
@narendly, @dasahcc. 

This need to be validated.

Reply:
Does this need to be thread-safe? This data is statically populated at the very beginning and no one is updating this data, right (update not allowed)?

Reply:
_metadataStoreRoutingData.getMetadataStoreRealm for sure should be thread safe. All the public API in this one should be as they will be called by multiple thread. 

We probably should raise this point in the diff implementing this code. I just forget to look careful there. 

That said, it is not hard to make it thread safe as you said, think they are immutable once published. Wrapping the data to be published in a immutable collection seems to be the easiest way to me.

Reply:
Could you look at how _metadataStoreRoutingData is implemented? It sounds like you think it's returning a map - it's actually an interface that does not provide any APIs that can modify the underlying data structure.